src/webgpu/shader/types.ts

```diff
@@ -36,3 +42,146 @@ export const kMatrixContainerTypeInfo = /* prettier-ignore */ {
 } as const;
 /** List of all matNxN<> container types. */
 export const kMatrixContainerTypes = keysOf(kMatrixContainerTypeInfo);
+
+/**
+ * Generate a bunch types (vec, mat, sized/unsized array) for testing.
+ */
+export function* generateTypes({
+  storageClass,
+  baseType,
+  containerType,
+  supportsAtomics = false,
+}: {
+  storageClass: string;
+  /** Base scalar type (i32/u32/f32/bool). */
+  baseType: ScalarType;
+  /** Container type (scalar/vector/matrix/array) */
+  containerType: ContainerType;
+  /** Whether to wrap the baseType in `atomic<>`. */
+  supportsAtomics?: boolean;
+}) {
+  const scalarInfo = kScalarTypeInfo[baseType];
+  if (supportsAtomics) {
+    assert(scalarInfo.supportsAtomics, 'type does not support atomics');
+  }
+  const scalarType = supportsAtomics ? `atomic<${baseType}>` : baseType;
+
+  // Storage and uniform require host-sharable types.
+  if (storageClass === 'storage' || storageClass === 'uniform') {
+    assert(isHostSharable(baseType), 'type ' + baseType.toString() + ' is not host sharable');
+  }
+
+  // Scalar types
+  if (containerType === 'scalar') {
+    yield {
+      type: `${scalarType}`,
+      _kTypeInfo: {
+        elementBaseType: `${scalarType}`,
```

Comment:
what if elementBaseType were a full _kTypeInfo as well? could be useful if/when we have nested containers

Reply:
Let's leave this as-is. I'll add this utility in my next CL.
Although I think this would only benefit arrays. Since other baseTypes has to be a ScalarType. ie. vec4x4<vec2x2<f32>> is invalid.